File: wpinet/src/test/native/cpp/UsbDeviceDetectorTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/net/UsbDeviceDetector.hpp"

#include <catch2/catch_test_macros.hpp>

#include "UsbDeviceDetectorImpl.hpp"

namespace wpi::net {

TEST_CASE("UsbDeviceDetector validates port", "[usb]") {
  REQUIRE_NOTHROW(UsbDeviceDetector{0, {}, {}});
  REQUIRE_NOTHROW(UsbDeviceDetector{3, {}, {}});
  REQUIRE_THROWS_AS(UsbDeviceDetector(-1, {}, {}), std::out_of_range);
  REQUIRE_THROWS_AS(UsbDeviceDetector(4, {}, {}), std::out_of_range);
}

TEST_CASE("SystemCore USB topology maps to physical ports", "[usb]") {
  REQUIRE(detail::GetSystemCoreUsbPort("3-1.1") == 0);
  REQUIRE(detail::GetSystemCoreUsbPort("3-1.2") == 1);
  REQUIRE(detail::GetSystemCoreUsbPort("3-1.3") == 2);
  REQUIRE(detail::GetSystemCoreUsbPort("3-1.4") == 3);
  REQUIRE(detail::GetSystemCoreUsbPort("3-1.1.2") == 0);

  REQUIRE_FALSE(detail::GetSystemCoreUsbPort("3-1"));
  REQUIRE_FALSE(detail::GetSystemCoreUsbPort("3-1.0"));
  REQUIRE_FALSE(detail::GetSystemCoreUsbPort("3-1.5"));
  REQUIRE_FALSE(detail::GetSystemCoreUsbPort("3-1.10"));
  REQUIRE_FALSE(detail::GetSystemCoreUsbPort("2-1.1"));
}

}  // namespace wpi::net

File: wpinet/src/test/java/org/wpilib/net/UsbDeviceDetectorTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.net;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class UsbDeviceDetectorTest {
  @Test
  void validatesPort() {
    assertThrows(
        IllegalArgumentException.class,
        () -> new UsbDeviceDetector(-1, device -> {}, device -> {}));
    assertThrows(
        IllegalArgumentException.class, () -> new UsbDeviceDetector(4, device -> {}, device -> {}));
  }

  @Test
  void desktopHasNoImplementation() {
    try (UsbDeviceDetector detector = new UsbDeviceDetector(0, device -> {}, device -> {})) {
      assertFalse(detector.hasImplementation());
      assertFalse(detector.start());
      assertFalse(detector.isRunning());
      detector.stop();
    }
  }
}
